packages/kokkos/core/src/impl/Kokkos_Core.cpp

```diff
@@ -93,21 +93,80 @@ setenv("MEMKIND_HBW_NODES", "1", 0);
   // if the exact device is not set, but ndevices was given, assign round-robin using on-node MPI rank
-  if (use_gpu < 0 && ndevices >= 0) {
+  if (use_gpu < 0) {
     auto local_rank_str = std::getenv("OMPI_COMM_WORLD_LOCAL_RANK"); //OpenMPI
     if (!local_rank_str) local_rank_str = std::getenv("MV2_COMM_WORLD_LOCAL_RANK"); //MVAPICH2
     if (!local_rank_str) local_rank_str = std::getenv("SLURM_LOCALID"); //SLURM
-    if (local_rank_str) {
+
+    auto ctest_process_count_str = std::getenv("CTEST_PROCESS_COUNT"); //CTest
+    if (!ctest_device_type.empty() && ctest_process_count_str && local_rank_str) {
+      auto process_count = std::atoi(ctest_process_count_str);
       auto local_rank = std::atoi(local_rank_str);
-      use_gpu = local_rank % ndevices;
-    } else {
-      // user only gave us ndevices, but the MPI environment variable wasn't set.
-      // start with GPU 0 at this point
-      use_gpu = 0;
+
+      if (process_count > local_rank) {
+        std::ostringstream ctest_process;
+        ctest_process << "CTEST_PROCESS_" << local_rank;
+        auto ctest_process_str = std::getenv(ctest_process.str().c_str());
+
+        if (ctest_process_str) {
+          bool found_device = false;
+          auto begin = ctest_process_str;
+          auto comma = std::strchr(begin, ',');
+          while (comma) {
+            if (!std::strncmp(begin, ctest_device_type.c_str(), comma - begin)) {
+              found_device = true;
+              break;
+            }
+            begin = comma + 1;
+            comma = std::strchr(begin, ',');
+          }
+
+          if (!found_device && !strcmp(begin, ctest_device_type.c_str())) {
+            found_device = true;
+          }
+
+          if (found_device) {
+            std::string ctest_device_type_upper = ctest_device_type;
+            for (auto& c : ctest_device_type_upper) {
+              c = std::toupper(c);
+            }
+            ctest_process << "_" << ctest_device_type_upper;
+
+            auto resource_str = std::getenv(ctest_process.str().c_str());
+            if (resource_str) {
+              const char* colon = std::strchr(resource_str, ',');
+              if (colon && !strncmp(resource_str, "id:", 3)) {
+                std::string id(resource_str + 3, colon - resource_str - 3);
+                use_gpu = std::atoi(id.c_str());
+              } else {
+                use_gpu = 0;
+              }
+            } else {
+              use_gpu = 0;
+            }
+          } else {
+            use_gpu = 0;
+          }
+        } else {
+          use_gpu = 0;
+        }
+      } else {
+        use_gpu = 0;
+      }
+    } else if (ndevices >= 0) {
+      if (local_rank_str) {
+        auto local_rank = std::atoi(local_rank_str);
+        use_gpu = local_rank % ndevices;
+      } else {
+        // user only gave us ndevices, but the MPI environment variable wasn't set.
+        // start with GPU 0 at this point
+        use_gpu = 0;
+      }
```

Comment:
That's a pretty long chunk of code.  It would be better to pull it out as a function and add unit tests for it.